File: src/main/com/intellij/lang/jsgraphql/schema/library/GraphQLLibrary.java
package com.intellij.lang.jsgraphql.schema.library;

import com.intellij.lang.jsgraphql.GraphQLBundle;
import com.intellij.lang.jsgraphql.icons.GraphQLIcons;
import com.intellij.navigation.ItemPresentation;
import com.intellij.openapi.roots.SyntheticLibrary;
import com.intellij.openapi.vfs.VirtualFile;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.Collection;
import java.util.Collections;
import java.util.Objects;

public class GraphQLLibrary extends SyntheticLibrary implements ItemPresentation {

  private final GraphQLLibraryDescriptor myLibraryDescriptor;
  private final VirtualFile myFile;

  public GraphQLLibrary(@NotNull GraphQLLibraryDescriptor libraryDescriptor,
                        @NotNull VirtualFile file) {
    myLibraryDescriptor = libraryDescriptor;
    myFile = file;
  }

  public @NotNull GraphQLLibraryDescriptor getLibraryDescriptor() {
    return myLibraryDescriptor;
  }

  @Override
  public @Nullable String getPresentableText() {
    return String.format("%s %s", GraphQLBundle.message("graphql.library.prefix"), myLibraryDescriptor.getPresentableText());
  }

  @Override
  public @Nullable Icon getIcon(boolean unused) {
    return GraphQLIcons.Logos.GraphQL;
  }

  @Override
  public @NotNull Collection<VirtualFile> getSourceRoots() {
    return Collections.singleton(myFile);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    GraphQLLibrary that = (GraphQLLibrary)o;
    return Objects.equals(myFile, that.myFile);
  }

  @Override
  public int hashCode() {
    return Objects.hash(myFile);
  }
}

File: src/main/com/intellij/lang/jsgraphql/schema/library/GraphQLLibrary.kt
package com.intellij.lang.jsgraphql.schema.library

import com.intellij.lang.jsgraphql.GraphQLBundle
import com.intellij.lang.jsgraphql.icons.GraphQLIcons
import com.intellij.navigation.ItemPresentation
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.platform.backend.workspace.virtualFile
import com.intellij.platform.workspace.storage.url.VirtualFileUrl
import javax.swing.Icon

data class GraphQLLibrary(
  val descriptor: GraphQLLibraryDescriptor,
  val rootUrls: Set<VirtualFileUrl>,
) : ItemPresentation {
  val sourceRoots: Collection<VirtualFile>
    get() = rootUrls.mapNotNull { url -> url.virtualFile?.takeIf { it.isValid } }.toSet()

  override fun getPresentableText(): String =
    GraphQLBundle.message("graphql.library.prefix", descriptor.displayName)

  override fun getIcon(unused: Boolean): Icon = GraphQLIcons.Logos.GraphQL
}
